Add documentation on new GCB functionality

ezimanyi opened the pull request.

ezimanyi committed:
Add documentation on new GCB functionality

dorbin commented on setup/ci/gcb.md:
```diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
 Google Cloud Build sends [Build Notifications](https://cloud.google.com/cloud-build/docs/send-build-notifications)
-when the state of your build changes. You can create a pipeline trigger to invoke a pipeline based
-on the status of your build.
+when the state of your build changes.  Spinnaker subscribes to these pub/sub messages so that in can track the status
```
```suggestion
when the state of your build changes.  Spinnaker subscribes to these pub/sub messages so that it can...
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
-when the state of your build changes. You can create a pipeline trigger to invoke a pipeline based
-on the status of your build.
+when the state of your build changes.  Spinnaker subscribes to these pub/sub messages so that in can track the status
+of builds it has initiated in a GCB stage and to trigger pipelines based on build status changes.
```
```suggestion
* track the status of builds it has initiated in a GCB stage
* trigger pipelines based on build-status changes
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
+assumed to be in a file at path `$SERVICE_ACCOUNT_KEY` in the below commands.
 
-2. Configure and deploy Spinnaker via Halyard with your newly created subscription.
+## Configure Spinnaker for Google Cloud Build Support
```
```suggestion
## Configure Spinnaker to work with Google Cloud Build
```

dorbin replied:
```suggestion
## Configure Spinnaker for Google Cloud Build support
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -8,7 +8,9 @@ sidebar:
+ * trigger pipelines when a GCB build completes
+ * add a GCB stage to your pipeline
 
 ## Prerequisites
```
So for prereqs, you need...
* a GCP project
* GCB API enabled (maybe we can keep these two combined in one bullet)
* subscription to cloud-build topic
* a service account that has both Cloud Build Editor and Pub/Sub Subscriber permissions

I think these should either be bullets, with content indented under them, or subheadings.  Subhead probably best.

ezimanyi replied:
Added subheadings for each

maggieneterval reviewed (approved):
Looks perfect!  :cloud: :hammer:

maggieneterval commented on setup/ci/gcb.md:
```diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
+of builds it has initiated in a GCB stage and to trigger pipelines based on build status changes.
 
-1. Create a Subscription object for the `cloud-builds` topic in your project:
+Create a Subscription object for the `cloud-builds` topic in your project:
```
Actually one random nit: should we either append a colon to all enumerated instructions that are followed by commands, or none of them?

ezimanyi replied:
Good point.  @dorbin , any preference or convention with regards to this?

dorbin replied:
I say include it. But if the introducing sentence trails off, such that the list item(s) complete the sentence, I like to use an elipsis...

ezimanyi replied:
Updated all instructions to be preceded by colons (as none seemed to fit the case for needing an elipsis)

dorbin commented on setup/ci/gcb.md:
````diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
       --project $PROJECT_ID
     ```
+    
+Finally, you will need a service account that has both Cloud Build Editor and Pub/Sub Subscriber permissions, which is
````
```suggestion
Finally, you will need a service account that has both Cloud Build Editor and Pub/Sub Subscriber permissions.
```

dorbin commented on setup/ci/gcb.md:
````diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
     ```
+    
+Finally, you will need a service account that has both Cloud Build Editor and Pub/Sub Subscriber permissions, which is
+assumed to be in a file at path `$SERVICE_ACCOUNT_KEY` in the below commands.
````
```suggestion
The commands below look for the service account key in a path/file defined in `$SERVICE_ACCOUNT_KEY`.
```

dorbin commented on setup/ci/gcb.md:
````diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
+```
 
 ## Configure Your Pipeline Trigger
 
````
While you're in here, can you change all these numbers to `1`? No reason to hardcode them.

ezimanyi replied:
Didn't realize you could use that syntax, thanks!

dorbin commented on setup/ci/gcb.md:
````diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
+```
 
 ## Configure Your Pipeline Trigger
 
````
Above, we're on the command line. Now, suddenly, Deck. Which is fine, but it might be a good idea to put a sentence here orienting the user. 
"Configure your pipeline to be triggered by a completed GCB build:"

dorbin commented on setup/ci/gcb.md:
````diff
@@ -20,35 +22,37 @@ You can enable the API with the following `gcloud` command:
-    ```
+```
 
 ## Configure Your Pipeline Trigger
````
```suggestion
## Configure your pipeline trigger
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
 appropriate values). You can then [use the produced image](/reference/artifacts/in-pipelines/)
 in downstream stages.
+
+## Configure a Google Cloud Build Stage
```
```suggestion
## Configure a Google Cloud Build stage
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
+
+## Configure a Google Cloud Build Stage
+
+To run a Google Cloud Build as part of a Spinnaker pipeline, create a stage of type *Google Cloud Build*.  Configure the
```
```suggestion
To run a GCB build as part of a Spinnaker pipeline, create a stage of type *Google Cloud Build*.  Configure the
```

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
+## Configure a Google Cloud Build Stage
+
+To run a Google Cloud Build as part of a Spinnaker pipeline, create a stage of type *Google Cloud Build*.  Configure the
+stage by selecting the GCB account to use to run the build, and entering the [build configuration YAML](https://cloud.google.com/cloud-build/docs/build-config)
```
I think all this should be broken out into steps.

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
 appropriate values). You can then [use the produced image](/reference/artifacts/in-pipelines/)
 in downstream stages.
+
+## Configure a Google Cloud Build Stage
```
I think this should be in /guides/pipeline/ rather than in setup. 
Everything from here on down.

ezimanyi replied:
I was wondering if that might be better for the stage instructions, but it looked like that section had mostly quick summaries rather than detailed instructions.  Let's chat about where this should live.

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
+the build logs in the Google Cloud Console:
+![](/setup/ci/gcb_status.png)
+
+## Configuration via Pub/Sub
```
The previous section was "Configure a Google Cloud Build stage." Is this section more like, "Configure a GCB trigger for your pipeline"?

ezimanyi replied:
It's really about the pre-1.14 config; changed the title to reflect that

dorbin commented on setup/ci/gcb.md:
```diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
+
+## Configuration via Pub/Sub
+
+Prior to version 1.14, Spinnaker did not have built-in support for Google Cloud Build, but pipelines could be
```
Now, as of 1.14, we do support GCB, but it's not equivalent to the former support described in this paragraph, right? Assuming I'm understanding this correctly. 
That is, this paragraph sounds like we're acknowledging the old way of doing something after saying the new way. And saying you can still do it the old way.
But aren't we doing two different things here:
* trigger from a GCBuild, via pub/sub
* spawn a GCBuild from a pipeline

If this understanding is correct, I'd change this paragraph to just introduce how to configure the trigger, removing references to before/after version number.

ezimanyi replied:
Spawning a GCB build from a pipeline is completely new, so there's no equivalent way of doing it prior to version 1.14.

Triggering from a GCB build could be done before via the method in this section (directly configuring pubsub) but going forward the recommended way is to configure as above.  I wanted to leave this section here for users who are not yet on 1.14, or who are upgrading to 1.14 but already have the old way configured.

Maybe it would help if I made it more clear that this is an alternate way of configuring triggering, but there's no way to start a GCB build from Spinnaker prior to 1.14?

dorbin commented on setup/ci/gcb.md:
````diff
@@ -73,3 +77,46 @@ if the build produces a Docker image, you can add an expected artifact of type *
+    hal deploy apply
+    ```
+
+The steps to create a pipeline trigger in this case are exactly the same except that the **Subscription Name** field
````
exactly the same as configuring the GCB stage? or same as what?

ezimanyi replied:
I'll change this to a link directly to the above instructions so it's clear

ezimanyi committed:
Apply suggestions from code review

Co-Authored-By: Dave Dorbin <ddorbin@google.com>

ezimanyi committed:
PR feedback

ezimanyi committed:
Next round of PR feedback

ezimanyi committed:
Merge branch 'master' into gcb-docs

ezimanyi commented:
I've gone through and addressed all the comments, leaving open the two comments that might need a further round of discussion.  As discussed, I'll merge this as is so it's live when we publish the 1.14 release notes, but can make a follow-up PR for any further changes.
